plugin-docker/che-plugin-docker-client/src/main/java/org/eclipse/che/plugin/docker/client/JsonMessageReader.java

```diff
@@ -0,0 +1,70 @@
+/*******************************************************************************
+ * Copyright (c) 2012-2015 Codenvy, S.A.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Codenvy, S.A. - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.che.plugin.docker.client;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonIOException;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonStreamParser;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.PushbackInputStream;
+
+/**
+ * Docker daemon sends chunked data in response. One chunk isn't always one JSON object so need to read full chunk at once to be able
+ * restore JSON object. This reader merges (if needs) few chunks until get full JSON object that we can parse.
```

Comment:
Full JSON here means valid JSON. Chunk may end with not valid JSON. To make data valid JSON it is needed to read next chunk, concatenate it with the previous one and then try to parse into JSON. 
By the way I did not change the logic of JSON parsing, just moved code to this class and made parsing parameterized.
Authors of this code are Andrew and you. And comment was written by someone of you.